File: error_model.py
__author__ = 'Gongyu'

import edit_distance


def split_error(word, key, context_num=2):
    re_list = []
    dis, pos_vector = edit_distance.edit(word, key)

    for pos in xrange(len(pos_vector)):
        if pos_vector[pos][0] != pos_vector[pos][1]:
            re_list.append(pos_vector[pos][0] + '|' + pos_vector[pos][1])

            for slide in xrange(1, context_num + 1):
                head = max(0, pos - slide)
                for i in xrange(head, pos + 1):
                    if i + slide >= len(pos_vector):
                        continue
                    slice_w = ''
                    slice_s = ''
                    for ii in xrange(slide + 1):
                        if pos_vector[i + ii][0] != '#':
                            slice_w += pos_vector[i + ii][0]
                        if pos_vector[i + ii][1] != '#':
                            slice_s += pos_vector[i + ii][1]
                    re_list.append(slice_w + '|' + slice_s)
    return re_list


def dump_error_data(train_file, error_file, context_num=2):
    infile = open(train_file, 'r')
    outfile = open(error_file, 'w')
    error_map = {}

    print 'dump error data...'
    # error_list = error_map.items()
    # error_list.sort(key=lambda item: item[1])
    for key in error_map.keys():
        outfile.write(key + '\t' + str(error_map[key]) + '\n')
        outfile.flush()

if __name__ == '__main__':
    dump_error_data('training_data.txt', 'error_data.txt')
    # print split_error('you', 'yiu')

File: src/cootek/spell/model/ErrorModel.java
package cootek.spell.model;

import cootek.spell.bean.Pair;
import cootek.spell.tool.EditDistance;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * Created by gongyu on 2015/4/25.
 */
public class ErrorModel {

    public List<String> splitError(String word, String source, int contex_num){
        List<String> split_list = new ArrayList<>();
        List<Pair<Character>> pos_list = new ArrayList<>();
        new EditDistance().edit(word, source, pos_list);
        // System.out.println(pos_list);

        for(int pos = 0; pos < pos_list.size(); pos++){
            if(pos_list.get(pos).val1 == pos_list.get(pos).val2) continue;

            split_list.add(pos_list.get(pos).val1 + " | " + pos_list.get(pos).val2);

            for(int slide = 1; slide <= contex_num; slide++){
                int head = Math.max(0, pos - slide);
                for(int i = head; i <= pos; i++){
                    if(i + slide >= pos_list.size()) continue;
                    StringBuffer slice_w = new StringBuffer("");
                    StringBuffer slice_s = new StringBuffer("");

                    for(int ii = 0; ii <= slide; ii++){
                        if(pos_list.get(i+ii).val1 != '#') slice_w.append(pos_list.get(i+ii).val1);
                        if(pos_list.get(i+ii).val2 != '#') slice_s.append(pos_list.get(i+ii).val2);
                    }

                    split_list.add(slice_w.append(" | ").append(slice_s).toString());
                }
            }
        }
        return split_list;
    }

    public void dumpErrorData(String train_file, String error_file, int context_num) throws Exception{
        BufferedReader br = new BufferedReader(new FileReader(train_file));
        BufferedWriter bw = new BufferedWriter(new FileWriter(error_file));

        System.out.println("read and split error data ...");
        HashMap<String, Integer> error_map = new HashMap<>();
        String line;
        while((line = br.readLine()) != null) {
            if(line.trim().equals("")) continue;

            String[] strs = line.trim().split("\t");
            if(strs[0].equals(strs[1])) continue;

            // System.out.println(strs[0] + "\t" + strs[1]);
            List<String> split_list = splitError(strs[0], strs[1], context_num);
            for(String error : split_list){
                if(!error_map.containsKey(error)) error_map.put(error, Integer.parseInt(strs[2]));
                else error_map.put(error, error_map.get(error) + Integer.parseInt(strs[2]));
            }
        }

        System.out.println("dump error data ...");
        for(String key : error_map.keySet()){
            bw.write(key + "\t" + error_map.get(key));
            bw.newLine();
            bw.flush();
        }

        br.close();
        bw.close();
    }

    public static void main(String[] args) throws Exception{
        System.out.println(new cootek.spell.model.ErrorModel().splitError("prepare", "perpare", 2));
        // new ErrorModel().dumpErrorData("training_data.txt", "error_data.txt", 2);
    }
}
